jetpack-io/devbox
# [vscode-extension] reopen vscode in devbox env using process communication

## Summary
Added ability to reload a vscode window and re-open it in a devbox environment.
This is done in 4 steps: 
1. Devbox extension invokes devbox CLI (devbox integrate vscode) as a child process.
2. CLI then prepares the devbox environment (computeNixEnv) and sends a message to parent process that preparing new environment is finished and ready.
3. Parent process (vscode window) upon receiving the finish message, closes itself. 
4. CLI will open a new vscode window (cmd.execute) with env variables it has received from setting up devbox environment.

NOTE: This required changes to both CLI and the extension, so the release of extension has to happen after CLI.

Addresses #1029 

## How was it tested?
1. compile CLI and put the binary in an directory with devbox.json (let's call it `vstest/`)
2. specify a package (e.g., hello@latest) and an env variable in devbox.json (e.g., "foo": "bar123")
3. in devbox.ts update `const devbox = 'devbox';` to `const devbox = <absolute path to compiled CLI binary>`
4. `cd devbox/vscode-extension` then package an installable extensinon file: `vsce package`.(this will create a devbox-version.vsix file)
5. install this extension on vscode by `code --install-extension path-to-vsix-file`
6. open `vstest/` in vscode. 
7. In vscode's terminal run `echo $foo` it should be empty. Then type `which hello` and it should point to devbox environment's hello, or say command not found
8. open command palette in vscode (cmd + shift + p) and type "reopen in devbox" and choose the suggested action.
9. once vscode is re-opened, try step 7 and confirm that vscode now has devbox environment setup.
